tools_lib/src/stack_collapse.rs

```diff
@@ -1,183 +1,215 @@
+        }
 
-            while event.timestamp > *next_observation_time {
-                *count += 1;
-                *next_observation_time += Duration::from_millis(interval);
+        let sampling_interval_nanos = sampling_interval.as_nanos();
+
+        let thread_duration_nanos = max_sampled_timestamp.duration_since(thread_start_time).unwrap().as_nanos();
+        let samples = thread_duration_nanos / sampling_interval_nanos;
+        thread_start_time + Duration::from_nanos((samples * sampling_interval_nanos).try_into().unwrap())
+    } else {
+        // No events? Ignore the thread.
+        return
+    };
+
+    let mut stack: Vec<Event<'_>> = vec![];
+    // The id is updated in sync we `stack`
+    let mut stack_id = "rustc".to_owned();
+
+    loop {
+
+        // Pop things from the stack that we've moved past
+        while let Some(top) = stack.last().cloned() {
+            if top.timestamp.contains(sample_cursor) {
+                break
+            } else {
+                let popped = stack.pop().unwrap();
+                let new_stack_id_len = stack_id.len() - (popped.label.len() + 1);
+                stack_id.truncate(new_stack_id_len);
             }
         }
 
-        match event.timestamp_kind {
-            TimestampKind::Start => {
-                thread_stack.push(event);
+        // Push things onto the stack that are under the cursor now
+        while let Some(event) = events.peek().cloned() {
+            if sample_cursor >= event.timestamp.end() {
+                // The next event is not under the cursor, so break
+                break
             }
-            TimestampKind::End => {
-                let previous_event = thread_stack.pop().expect("no start event found");
-                assert_eq!(event.label, previous_event.label);
-                assert_eq!(previous_event.timestamp_kind, TimestampKind::Start);
+
+            // Check if the event is under the cursor. It could also be an
+            // event that the cursor has skipped entirely
+            if event.timestamp.contains(sample_cursor) {
+                stack_id.push(';');
+                stack_id.push_str(&event.label[..]);
+                stack.push(event);
             }
-            TimestampKind::Instant => {}
+
+            events.next();
         }
+
+        if events.peek().is_none() && stack.is_empty() {
+            break
+        }
+
+        eprintln!("{}", stack_id);
+        *counters.entry(stack_id.clone()).or_insert(0) += 1;
+
+        sample_cursor -= sampling_interval;
     }
+}
 
-    recorded_stacks
+pub fn collapse_stacks<'a>(
+    profiling_data: &ProfilingData,
+    sampling_interval: Duration,
+) -> HashMap<String, u64> {
+    let thread_ids = find_all_thread_ids_and_start_times(profiling_data);
+    let mut counters = HashMap::new();
+
+    for (thread_id, thread_start_time) in thread_ids {
+        process_thread(thread_id, thread_start_time, profiling_data, sampling_interval, &mut counters);
+    }
```

Comment:
This implementation iterators the profiling data once for each thread and once in the beginning for finding all thread ids. It think this makes the code quite a bit simpler. An alternative approach that is exactly as simple but trades processing time for memory consumption is to iterate once and building up a vec of events for each thread.

Reply:
a bit worried about the time for e.g when compiling servo script and using incremental there will be around 1500 threads and to iterate that many times over all events feels like it will take forever